Add Date aggregation {wip} (#1585)
enkidevs/curriculum: mongodb/intro/date-aggregation/aggregation-with-date-expression-operators-v.md

---
author: Stefan-Stojanovic

aspects:
  - introduction

type: normal

category: must-know

> links:
> - '[Date Operators](https://docs.mongodb.com/manual/reference/operator/aggregation/#date-expression-operators){documentation}'

Reviewer: ```suggestion
links:
  - '[Date Operators](https://docs.mongodb.com/manual/reference/operator/aggregation/#date-expression-operators){documentation}'
```

---
# Aggregation With All Date Operators
---
## Content

Now that we've seen all the date operators and know how to use them, let's take a look at an example of all operators being used in the same aggregation.

> **Note:** We are using the same document[1] as in the previous insights.

Reviewer: ```suggestion
> 💡 We are using the same document[1] as in the previous insights.
```

Example with all operators:
```js
db.pokemon.aggregate([
  { $match: { name: "Bulbasaur" } },
  {
    $project: {
      dayOfTheYear: { $dayOfYear: "$createdAt" },
      createdAtYear: { $year: "$createdAt" },
      createdAtMonth: { $month: "$createdAt" },
      dayOfTheMonth: { $dayOfMonth: "$createdAt" },
      createdAtHour: { $hour: "$createdAt" },
      createdAtMinute: { $minute: "$createdAt" },
      createdAtSecond: { $second: "$createdAt" },
      createdAtMillisecond: { $millisecond: "$createdAt" },
      weekOfTheYear: { $week: "$createdAt" },
      dayOfTheWeek: { $dayOfWeek: "$createdAt" },
      _id: 0
    }
  }
]);
```

Output:
```js
{ 
  "dayOfTheYear": 282, 
  "createdAtYear": 2019, 
  "createdAtMonth": 10, 
  "dayOfTheMonth": 9, 
  "createdAtHour": 7, 
  "createdAtMinute": 21, 
  "createdAtSecond": 14, 
  "createdAtMillisecond": 0, 
  "weekOfTheYear": 40, 
  "dayOfTheWeek": 4 
}

> ```

Reviewer: can we use a more creative example? imo there is no benefit in this example because we've already shown this in all the previous insights. all that it does is add all the operators in the project stage

we could do something like finding the remaining time until one's birthday, the number of days remaining untill a contract renewal or something else. also, we don't have to include all the operators as long as we use an example that shows how a user could use these operators in a real-life application

edit: I notice that you give some examples in the first insight. why not make use of those? "These operators can be used to extract the specific time at which a document was made, they can be used to find a group of documents created on a specific date or to find all documents between 2 dates."

---

> ## Practice

Reviewer: we should add a revision question for this as well

---
author: Stefan-Stojanovic

levels:
  - beginner
  
aspects:
  - introduction

type: normal

category: must-know

links:

---

> # Aggregation with all 10 date expression operators

Reviewer: ```suggestion
# Aggregation With All Date Operators
```

---
## Content

> Now that we've seen all 10 date operators and know which part of the date each is responsible for, here is an example of all operators used in the same aggregation.

Reviewer: ```suggestion
Now that we've seen all the date operators and know how to use them, let's take a look at an example of all operators being used in the same aggregation.
```

> **Note:** Same document as in the previous insights. (check footnotes Previous Document[1])

Reviewer: ```suggestion
**Note:** We are using the same document[1] as in the previous insights.
```

Example with all operators:
```js
db.pokemon.aggregate(
   [
     {$match: {"name": "Bulbasaur"}},
     {
       $project:
         {
           dayOfTheYear: { $dayOfYear: "$Date" },
           Year: { $year: "$Date" },
           Month: { $month: "$Date" },
           DayOfTheMonth: { $dayOfMonth: "$Date" },
           Hour: { $hour: "$Date" },
           Minutes: { $minute: "$Date" },
           Seconds: { $second: "$Date" },
           Milliseconds: { $millisecond: "$Date" },
           Week: { $week: "$Date" },
           DayOfTheWeek: { $dayOfWeek: "$Date" },
           _id: 0
         }
     }
   ]

> )

Reviewer: put the code blocks in this insight through prettier as well

```

Output:
```js
{ 
  "dayOfTheYear": 282, 
  "Year": 2019, 
  "Month": 10, 
  "DayOfTheMonth": 9, 
  "Hour": 7, 
  "Minutes": 21, 
  "Seconds": 14, 
  "Milliseconds": 0, 
  "Week": 40, 
  "DayOfTheWeek": 4 
}
```

---
## Practice

> Fill in the blanks bellow to get all parts of the date the document with a `name`: `"Bulbasaur"` was created on.

Reviewer: ```suggestion
Fill in the blanks below to get all parts of the date the document with `name: "Bulbasaur"` was created on.
```

```js
db.pokemon.aggregate(
   [
     {$match: {"name": "Bulbasaur"}},
     {
       $project:
         {
           dayOfTheYear: { ???: "$Date" },
           Year: { $year: "$Date" },
           Month: { $month: "$Date" },
           DayOfTheMonth: { $dayOfMonth: "$Date" },
           Hour: { $hour: "$Date" },
           Minutes: { $minute: "$Date" },
           Seconds: { $second: "$Date" },
           Milliseconds: { $millisecond: "$Date" },
           Week: { $week: "$Date" },
           DayOfTheWeek: { $dayOfWeek: "$Date" },
           _id: 0
         }
     }
   ]

> )

Reviewer: did you mean to leave only one `???` ? also, this question is a bit too easy as the field names are exactly the same as the stage names. maybe we can improve this question a bit?

---
author: Stefan-Stojanovic

levels:
  - beginner
  
aspects:
  - introduction

type: normal

category: must-know

links:

---
# Aggregation With All Date Operators
---
## Content

Now that we've seen all the date operators and know how to use them, let's take a look at an example of all operators being used in the same aggregation.

**Note:** We are using the same document[1] as in the previous insights.

Example with all operators:
```js
db.pokemon.aggregate([
  { $match: { name: "Bulbasaur" } },
  {
    $project: {

>       dayOfTheYear: { $dayOfYear: "$Date" },

Reviewer 2: this one is lowercase (as it's usually expected) but all others are capitalized

      Milliseconds: { $millisecond: "$Date" },
      Week: { $week: "$Date" },
      DayOfTheWeek: { $dayOfWeek: "$Date" },
      _id: 0
    }
  }
]);
```

Output:
```js
{ 
  "dayOfTheYear": 282, 
  "Year": 2019, 
  "Month": 10, 
  "DayOfTheMonth": 9, 
  "Hour": 7, 
  "Minutes": 21, 
  "Seconds": 14, 
  "Milliseconds": 0, 
  "Week": 40, 
  "DayOfTheWeek": 4 
}
```

---
## Practice

Fill in the blanks below to get all parts of the date the document with `name: "Bulbasaur"` was created on.
```js
db.pokemon.aggregate([
  { $match: { name: "Bulbasaur" } },
  {
    $project: {
      dayOfTheYear: { ???: "$Date" },
      Year: { ???: "$Date" },
      Month: { ???: "$Date" },
      DayOfTheMonth: { ???: "$Date" },
      Hour: { ???: "$Date" },
      Minutes: { ???: "$Date" },
      Seconds: { ???: "$Date" },
      Milliseconds: { ???: "$Date" },
      Week: { ???: "$Date" },
      DayOfTheWeek: { ???: "$Date" },
      _id: 0
    }
  }
]);
```

* `$dayOfYear`
* `$year`
* `$month`
* `$dayOfMonth`
* `$hour`
* `$minute`
* `$second`
* `$millisecond`
* `$week`
* `$dayOfWeek`
* `$dayOfTheYear`
* `$dayOfTheMonth`
* `$dayOfTheWeek`
* `$Year`
* `$Hour`
* `$Second`
* `$Minute`
* `$Week`

> * `$Month`

Reviewer: I think this question is still too easy, with the answers already being given in the question text. also, there are too many possible answers. can you make the question ask something like "Extract the day and the hour from the given date"

Author: Well, all the other insights ask for each part of the date. This is why this one, asks for all of them. We can add this question as a single RQ for the whole workout and leave this insight questionless?

Reviewer: hmm, yeah, you're right. let's just leave it as is.

---
## Footnotes

[1:Previous Document]
Here is the document used in the previous insight:
```javascript
// document without the Date
{ 
	"_id": ObjectId("5d9d8a6a0b24990f19398209"),
  "name": "Bulbasaur",
	"type": "Grass"
}
// Document with the Date
{ 
	"_id": ObjectId("5d9d8a6a0b24990f19398209"),
  "name": "Bulbasaur",
	"type": "Grass",
  "Date": ISODate("2019-10-09T07:21:14Z")
}

> ```

Reviewer: this also needs reformatting
